sklearn/multioutput.py

```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+        the predict method by indexing into the predictions with the
+        list of indices in the order attribute.
+
+
+    References
+    ----------
+    Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, "Classifier
+    Chains for Multi-label Classification", 2009.
+
+    """
+
+    def __init__(self, base_estimator, random_state=None, order=None, cv=None):
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        self.order = order
+        self.cv = cv
+
+    def fit(self, X, Y):
+        """Fit the model to data matrix X and targets Y.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : {array-like, sparse matrix}, shape (n_samples, n_classes)
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+
+        random_state = check_random_state(self.random_state)
+
+        if self.order is None:
+            self.order = list(range(Y.shape[1]))
+        elif self.order == 'random':
+            self.order = random_state.permutation(Y.shape[1])
+        elif sorted(self.order) != list(range(Y.shape[1])):
+                raise ValueError("invalid order")
+
+        self.estimators_ = [clone(self.base_estimator)
+                            for _ in range(Y.shape[1])]
+
+        self.classes_ = []
+
+        if sp.issparse(X):
+            if self.cv is None:
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    if chain_idx == 0:
+                        X_aug = X
+                    else:
+                        X_aug = sp.hstack((X, Y[:, self.order[:chain_idx]]))
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug, y)
+                    self.classes_.append(estimator.classes_)
+            else:
+                Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    if chain_idx == 0:
+                        X_aug = X
+                    else:
+                        X_aug = sp.hstack((X, Y[:, self.order[:chain_idx]]))
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug, y)
+                    Y_pred_chain[:, chain_idx] = cross_val_predict(
+                        self.base_estimator, X, y=y, cv=self.cv)
+                    self.classes_.append(estimator.classes_)
+
+        else:
+            if self.cv is None:
+                X_aug = np.hstack((X, Y[:, self.order]))
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug[:, :(X.shape[1] + chain_idx)], y)
+                    self.classes_.append(estimator.classes_)
+
+            else:
+                Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```

Comment:
You could again avoid doing the hstack in the loop by hstacking zeros onto X as above with X and Y